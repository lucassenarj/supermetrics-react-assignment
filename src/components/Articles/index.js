import React, { useEffect, useState } from "react";
import Article from "./../Article";
import InputText from "./../InputText";
import { FaSearch, FaArrowUp, FaArrowDown } from "react-icons/fa";
import { useSelector } from "react-redux";
import { sortPostsByParameter } from "./../../utils/helpers";
import "./index.scss";

function Articles() {
  const [posts, setPosts] = useState([]);
  const [active, setActive] = useState("ASC");
  const { articles } = useSelector(store => store.articles);
  
  useEffect(() => {
    setPosts(articles);
  }, [articles]);

  const handleSearch = ({ target: { value} }) => {
    if (value.trim() ===  "") {
      setPosts(articles);
      return;
    }

    const filtered = posts.filter(({ message }) => message.includes(value));
    setPosts(filtered);
  }

  const handleSort = (type) => {
    const sorted = sortPostsByParameter(posts, type);
    setPosts(sorted);
    setActive(type);
  };

  return (
    <div className="articles">
      <div className="articles__header">
        <div className="articles__buttons">
          <button
            onClick={() => handleSort("ASC")}
            className={["articles__buttons__btn", active === "ASC" ? "articles__buttons__btn--active" : ""].join(" ")}
            type="button"
            title="Most recent"
            data-testid="order-asc"
          >
            <FaArrowDown />
          </button>

          <button
            onClick={() => handleSort("DESC")}
            className={["articles__buttons__btn", active === "DESC" ? "articles__buttons__btn--active" : ""].join(" ")}
            type="button"
            title="Oldest"
            data-testid="order-desc"
            >
            <FaArrowUp />
          </button>
        </div>

        <div className="articles__search">
          <InputText
            icon={<FaSearch />}
            placeholder="Search Posts"
            type="search"
            onChange={handleSearch}
          />
        </div>
      </div>
      { posts.length === 0 && (<p>No article founded</p>) }
      { posts.map(({created_time, id, message}) => (
        <Article
          key={id}
          date={created_time}
          message={message}
        />))}
    </div>
  );
};

export default Articles;
