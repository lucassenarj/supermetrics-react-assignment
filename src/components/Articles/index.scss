@import "./../../assets/colors.scss";

.articles__header {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.articles__buttons {
  width: 120px;
  display: flex;
  justify-content: space-between;
}

.articles__buttons__btn {
  padding: 10px 20px;
  background-color: $button;
  cursor: pointer;
}

.articles__buttons__btn:hover {
  background-color: $button__hover;
}