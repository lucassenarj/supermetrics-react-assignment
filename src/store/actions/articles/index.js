export const ARTICLES_SET_LIST = "ARTICLES_SET_LIST";

export function setArticleList(posts) {
  return (dispatch) => {
    dispatch({
      type: ARTICLES_SET_LIST,
      payload: posts,
    });
  };
};
