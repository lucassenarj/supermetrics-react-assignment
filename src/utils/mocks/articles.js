export const articles = [
  {
    id: 'post60cf914a226ef_28c1e277',
    from_name: 'Carly Alvarez',
    from_id: 'user_6',
    message: 'try disability admiration patrol menu default shy civilian full borrow ecstasy race notice popular oak treaty',
    type: 'status',
    created_time: '2021-06-14T04:33:13+00:00'
  },
  {
    id: 'post60cf914a227b8_e716f0e6',
    from_name: 'Carly Alvarez',
    from_id: 'user_6',
    message: 'carbon faithful cigarette location contrast loud full proud confine sample building prevent calculation team pedestrian swell church relevance margin ministry evolution drill climate transport advice preparation corruption expertise treasurer extension handy electron pray brake extinct wreck mystery timber convince psychology herd area sweet ban access mine seed penny crew faithful tap publisher stubborn',
    type: 'status',
    created_time: '2021-06-08T16:27:02+00:00'
  },
  {
    id: 'post60cf914a22889_039dd6d8',
    from_name: 'Carly Alvarez',
    from_id: 'user_6',
    message: 'network fuel prestige hotdog terms line agile rhythm vague point lion tap toss size bend lion breast indoor underline clock tendency bay value tin fit realize crusade revolution dog racism vertical beer shorts initiative loose language reliance mile neighborhood roar stain rear fool deficiency rubbish fine ecstasy axis orchestra draw morning wine midnight lose restrict chaos sustain south recording train brick eyebrow say film franchise gift lamb drop title treasurer intermediate ton prisoner sailor rhythm printer',
    type: 'status',
    created_time: '2021-06-01T22:15:33+00:00'
  }
];
