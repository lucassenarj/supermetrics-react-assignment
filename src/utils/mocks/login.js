export const loginResponse = {
  email: "test@test.com",
  sl_token: "smslt_f893343e131e9_167c07f380c7305",
  expiration: "1624369506541",
  name: "User test",
};
