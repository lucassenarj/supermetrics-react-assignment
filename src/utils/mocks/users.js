export const users = [
  {
    user_id: 'user_4',
    user_name: 'Britany Heise',
    posts: [
      {
        id: 'post60cf914a225f1_37d9a0b4',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'conflict fight bend resident term fly coincidence put wagon west recruit faithful term rank coincidence guideline west coincide nuance roar race treasurer freedom cottage flow boat tired term definition smell forget facility final chest lip belong clearance printer qualified loose satellite say depression difficult point sailor dismissal leak safety',
        type: 'status',
        created_time: '2021-06-20T15:39:53+00:00'
      },
      {
        id: 'post60cf914a2262f_62849f6b',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'tick genuine racism leader prize retired suitcase dominant syndrome retired speed flawed beer book corn rehearsal critic union hiccup allocation variable belong beg snub midnight scheme accumulation penny tap',
        type: 'status',
        created_time: '2021-06-19T18:05:56+00:00'
      },
      {
        id: 'post60cf914a2269d_4856832f',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'smash suntan delicate irony cave superintendent speed satellite opposition fit thoughtful galaxy duck braid seed meadow habitat cheese product beer television giant competition barrier cluster trace mine bolt bike album loud spell resident write delay appear egg white patrol hand prisoner decorative desert braid anger pray deficiency tap',
        type: 'status',
        created_time: '2021-06-16T05:46:19+00:00'
      },
      {
        id: 'post60cf914a226e0_20f5688b',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'sword absorb awful bracket urine gradient pour plagiarize uncle adoption core makeup',
        type: 'status',
        created_time: '2021-06-14T14:10:05+00:00'
      },
      {
        id: 'post60cf914a226e5_1f523234',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'convention toss glory convince candle blade throne album reconcile terrify mess reliable multimedia smash linen agriculture bullet tin fool brave safety egg white introduce jurisdiction story jurisdiction velvet bend epicalyx format refuse superintendent solo boy ballot prescription abstract heal palm rescue safety mirror combine arch waist tie dressing sanctuary porter graze spend canvas printer computing scholar television second harmful',
        type: 'status',
        created_time: '2021-06-14T08:58:40+00:00'
      },
      {
        id: 'post60cf914a22747_d34228f9',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'noble dead index palm transmission aluminium pardon flourish herb wreck fly yard confront revoke introduce information speech spend jealous bolt rank cave relation swipe scream stain evening prisoner chocolate pedestrian crossing energy college upset gain conservative absent mystery slab menu drop',
        type: 'status',
        created_time: '2021-06-12T05:45:08+00:00'
      },
      {
        id: 'post60cf914a2277e_40aa13fb',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'good establish fill deadly executive',
        type: 'status',
        created_time: '2021-06-11T04:48:34+00:00'
      },
      {
        id: 'post60cf914a22816_1a4b4e6a',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'generation race talkative mile discourage dignity survey home mastermind conception debt bolt cluster bell water torture pool popular boy host vegetarian bother abstract broadcast tap palm glare blue jean horseshoe hate age final reinforce sailor cage kill parachute album convention field album trail federation tired survivor buy flower innocent environmental manage prize viable underline animal tile charm brick freckle psychology bar entitlement',
        type: 'status',
        created_time: '2021-06-05T17:59:55+00:00'
      },
      {
        id: 'post60cf914a2283f_43e890b1',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'pick axis cause achievement rule fame hotdog multimedia voyage line sacrifice transport dream delay patrol leave impulse dominant fame voyage sight format bill grip trouble lake run wisecrack crop addition tract tense entitlement chord loud ecstasy diamond handy district network flow carry officer baby franchise dressing peasant popular variant interest achievement benefit transport charm carve pick superintendent breed squash sensation bishop debate crusade rubbish evening',
        type: 'status',
        created_time: '2021-06-04T12:39:09+00:00'
      }
    ]
  },
  {
    user_id: 'user_6',
    user_name: 'Carly Alvarez',
    posts: [
      {
        id: 'post60cf914a226ef_28c1e277',
        from_name: 'Carly Alvarez',
        from_id: 'user_6',
        message: 'try disability admiration patrol menu default shy civilian full borrow ecstasy race notice popular oak treaty',
        type: 'status',
        created_time: '2021-06-14T04:33:13+00:00'
      },
      {
        id: 'post60cf914a227b8_e716f0e6',
        from_name: 'Carly Alvarez',
        from_id: 'user_6',
        message: 'carbon faithful cigarette location contrast loud full proud confine sample building prevent calculation team pedestrian swell church relevance margin ministry evolution drill climate transport advice preparation corruption expertise treasurer extension handy electron pray brake extinct wreck mystery timber convince psychology herd area sweet ban access mine seed penny crew faithful tap publisher stubborn',
        type: 'status',
        created_time: '2021-06-08T16:27:02+00:00'
      },
      {
        id: 'post60cf914a22889_039dd6d8',
        from_name: 'Carly Alvarez',
        from_id: 'user_6',
        message: 'network fuel prestige hotdog terms line agile rhythm vague point lion tap toss size bend lion breast indoor underline clock tendency bay value tin fit realize crusade revolution dog racism vertical beer shorts initiative loose language reliance mile neighborhood roar stain rear fool deficiency rubbish fine ecstasy axis orchestra draw morning wine midnight lose restrict chaos sustain south recording train brick eyebrow say film franchise gift lamb drop title treasurer intermediate ton prisoner sailor rhythm printer',
        type: 'status',
        created_time: '2021-06-01T22:15:33+00:00'
      }
    ]
  },
];