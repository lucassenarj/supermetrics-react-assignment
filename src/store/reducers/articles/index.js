import { ARTICLES_SET_LIST } from "./../../actions";

const initialState = {
  articles: [],
};

const articles = (state = initialState, action) => {
  switch(action.type) {
    case ARTICLES_SET_LIST: {
      const articles = action.payload;
      return {
        ...state,
        articles,
      }
    }
    default: 
      return state;
  }
}

export default articles;
