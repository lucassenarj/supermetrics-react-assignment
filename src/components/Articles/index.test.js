import * as React from "react";
import configureStore from "redux-mock-store";
import { render, fireEvent } from "@testing-library/react";
import * as redux from "react-redux";
import { articles } from "./../../utils/mocks/articles";
import Articles from "./";

const mockStore = configureStore([]);

describe("Components", () => {
  describe("Articles", () => {
    let wrapper;
    let store;
    const dispatch = jest.fn();
    const handleSort = jest.fn();
  
    beforeEach(() => {
      store = mockStore({
        articles: {
          articles
        }
      });

      jest.spyOn(redux, "useDispatch").mockImplementation(() => dispatch);

      wrapper = render(
        <redux.Provider store={store}>
          <Articles onClick={handleSort} />
        </redux.Provider>
      );
    });
    it("renders correctly", () => {
      const { container } = wrapper;
      expect(container.firstChild).toMatchSnapshot();
    });

    it("should click button order-asc button", async () => {
      const { getByTestId } = wrapper;
      const button = getByTestId("order-asc");
      fireEvent.click(button);

      expect(button).toHaveClass("articles__buttons__btn--active");
    });

    it("should click button order-desc button", () => {
      const { getByTestId } = wrapper;
      const button = getByTestId("order-desc");
      fireEvent.click(button);

      expect(button).toHaveClass("articles__buttons__btn--active");
    });
  });
});
