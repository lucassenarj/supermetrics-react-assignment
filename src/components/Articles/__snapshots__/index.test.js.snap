// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Articles renders correctly 1`] = `
<div
  class="articles"
>
  <div
    class="articles__header"
  >
    <div
      class="articles__buttons"
    >
      <button
        class="articles__buttons__btn articles__buttons__btn--active"
        data-testid="order-asc"
        title="Most recent"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
          />
        </svg>
      </button>
      <button
        class="articles__buttons__btn "
        data-testid="order-desc"
        title="Oldest"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
          />
        </svg>
      </button>
    </div>
    <div
      class="articles__search"
    >
      <div
        class="input-text"
      >
        
        <div
          class="input-text__formControl"
        >
          <span>
             
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
          </span>
          <input
            class="input-text__formControl__input"
            placeholder="Search Posts"
            type="search"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <article>
    <div
      class="article__header"
    >
      <p>
        June 14, 2021, 04:33:13
      </p>
    </div>
    <div
      class="article__content"
    >
      <p>
        try disability admiration patrol menu default shy civilian full borrow ecstasy race notice popular oak treaty
      </p>
    </div>
  </article>
  <article>
    <div
      class="article__header"
    >
      <p>
        June 8, 2021, 16:27:02
      </p>
    </div>
    <div
      class="article__content"
    >
      <p>
        carbon faithful cigarette location contrast loud full proud confine sample building prevent calculation team pedestrian swell church relevance margin ministry evolution drill climate transport advice preparation corruption expertise treasurer extension handy electron pray brake extinct wreck mystery timber convince psychology herd area sweet ban access mine seed penny crew faithful tap publisher stubborn
      </p>
    </div>
  </article>
  <article>
    <div
      class="article__header"
    >
      <p>
        June 1, 2021, 22:15:33
      </p>
    </div>
    <div
      class="article__content"
    >
      <p>
        network fuel prestige hotdog terms line agile rhythm vague point lion tap toss size bend lion breast indoor underline clock tendency bay value tin fit realize crusade revolution dog racism vertical beer shorts initiative loose language reliance mile neighborhood roar stain rear fool deficiency rubbish fine ecstasy axis orchestra draw morning wine midnight lose restrict chaos sustain south recording train brick eyebrow say film franchise gift lamb drop title treasurer intermediate ton prisoner sailor rhythm printer
      </p>
    </div>
  </article>
</div>
`;
